rcljava/src/main/java/org/ros2/rcljava/events/EventImpl.java

```diff
@@ -0,0 +1,164 @@
+   *     that created this event.
+   * @param handle A pointer to the underlying ROS 2 event structure, as an integer.
+   *     Must not be zero.
+   * @param messageType The <code>Class</code> of the messages that this
+   *     subscription will receive. We need this because of Java's type erasure,
+   *     which doesn't allow us to use the generic parameter of
+   *     @{link org.ros2.rcljava.Subscription} directly.
+   * @param callback The callback function that will be called when the event
+   *     is triggered.
+   */
+  public EventImpl(
+      final Class<ParentT> parentType,
+      final WeakReference<ParentT> parentReference,
+      final long handle,
+      final Class<T> eventStatusType,
+      final Consumer<T> callback) {
+    this.parentType = parentType;
+    this.parentReference = parentReference;
+    this.handle = handle;
+    this.eventStatusType = eventStatusType;
+    this.callback = callback;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public final Class<T> getStatusEventType() {
+    return this.eventStatusType;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public final Class<ParentT> getParentType() {
+    return this.parentType;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public final WeakReference<ParentT> getParentReference() {
+    return this.parentReference;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public final long getHandle() {
+    return this.handle;
+  }
+
+  /**
+   * Destroy a ROS 2 event (rcl_event_t).
+   *
+   * @param handle A pointer to the underlying ROS 2 event structure,
+   *     as an integer. Must not be zero.
+   */
+  private static native void nativeDispose(long handle);
+
+  /**
+   * {@inheritDoc}
+   */
+  public final void dispose() {
+    nativeDispose(this.handle);
+    this.handle = 0;
+  }
+
+  /**
+   * Takes the RCL event status and returns a pointer to it.
+   *
+   * @param event_handle A pointer to the underlying ROS 2 event (rcl_event_t).
+   * @param event_status_handle A pointer to the underlying ROS 2 event status (void *).
+   *     Must not be zero.
+   */
+  private static native void nativeTake(long event_handle, long event_status_handle);
+
+  /**
+   * {@inheritDoc}
+   */
+  public final void executeCallback() {
+    EventStatus eventStatus = null;
+    try {
+        eventStatus = eventStatusType.getDeclaredConstructor().newInstance();
+    } catch (NoSuchMethodException nme) {
+        nme.printStackTrace();
+    } catch (InvocationTargetException ite) {
+        ite.printStackTrace();
+    } catch (InstantiationException ie) {
+        ie.printStackTrace();
+    } catch (IllegalAccessException iae) {
+        iae.printStackTrace();
+    }
+    long nativeEventStatusHandle = eventStatus.allocateRCLStatusEvent();
+    nativeTake(this.handle, nativeEventStatusHandle);
+    eventStatus.fromRCLEvent(nativeEventStatusHandle);
+    eventStatus.deallocateRCLStatusEvent(nativeEventStatusHandle);
```

Comment:
Should we call `this.callback` here somewhere?

Reply:
:joy: yes

Reply:
Done in https://github.com/osrf/ros2_java/pull/4/commits/23480443ddb90f34d86de41b6b45529841a54269.